### PR TITLE
Create hal nrf 802154 repo

### PR DESCRIPTION
**Pull request by pkwiek**

### CODE REVIEW

**Comment by pkwiek** [comment]
If you have some spare time, please look at structure of the new repository. Do you have any thought regarding possible beneficial changes or simplifications as all the open-source modules are in one place?

**Review by jciupis — CHANGES_REQUESTED** [review]
I have several comments regarding the repository structure:
* since the repository name contains `nrf-802154` suffix, I would remove `nrf_802154_` prefixes from directory names. `nrf_802154_serialization` would then become `serialization` and `nrf_802154_sl` simply `sl` or alternatively `service_layer`, `service_library`.
* I would keep symmetry between serialization, SL and driver in the structure. Right now the driver resides in the main directory along with directories that contain the remaining components. In my opinion it would be clearer to keep the structure similar to the one present in hal_nordic repository so far, with all the components located at the same depth.
* There's no `lib` directory inside SL. Also, I'm wondering if it makes sense to keep a subdirectory `open` for sources of the opensource SL since no other sources are expected to ever be there.

**Comment by pkwiek** [comment]
>     * since the repository name contains `nrf-802154` suffix, I would remove `nrf_802154_` prefixes from directory names. `nrf_802154_serialization` would then become `serialization` and `nrf_802154_sl` simply `sl` or alternatively `service_layer`, `service_library`.
I am all for it.

>     * I would keep symmetry between serialization, SL and driver in the structure. Right now the driver resides in the main directory along with directories that contain the remaining components. In my opinion it would be clearer to keep the structure similar to the one present in hal_nordic repository so far, with all the components located at the same depth.
Directory structure is taken from https://github.com/NordicSemiconductor/nRF-IEEE-802.15.4-radio-driver, however I like solution proposed by you. This compartmentalization makes clearer what every part of this repo is for. As for now, there are source files in directories other than top-level `src` and that seems messy IMO.

>     * There's no `lib` directory inside SL. Also, I'm wondering if it makes sense to keep a subdirectory `open` for sources of the opensource SL since no other sources are expected to ever be there.
This repository will not contain libraries as it place is in zephyrproject-rtos, hence no directory. As for `open` directory I am unsure about its purpose as well.

**Comment by hubertmis** [comment]
I like @jciupis proposal. I would like to discuss a little deeper the last one.

> > ```
> > * There's no `lib` directory inside SL. Also, I'm wondering if it makes sense to keep a subdirectory `open` for sources of the opensource SL since no other sources are expected to ever be there.
> > ```
> 
> This repository will not contain libraries as it place is in zephyrproject-rtos, hence no directory. As for `open` directory I am unsure about its purpose as well.

Please keep it mind, that we expect this repository as a mirror of sdk-802154 with one directory (sl/lib) removed. In sdk-802154 directory structure would be possibly like:
```
sdk-802154/
├── driver
│   ├── include  // public API headers
│   └── src  // source and private headers
├── serialization
│   ├── include
│   ├── platform  // bare-metal examples, not zephyr
│   └── src
└── sl
    ├── include
    ├── lib
    └── platform // bare-metal examples, not zephyr
```

In this structure we should find a place for sources of open-source sl variant. If we created `sl/src/`, I would assume that there are source files used to build libraries in `sl/lib/`. To avoid such confusion, I proposed to create `open` subdirectory like `sl/open/src` and place open-SL sources there.

What do you think about it?

**Comment by jciupis** [comment]
@hubertmis That seems perfect to me.

**Comment by hubertmis** [comment]
Shall we target something like this?
```
sdk-802154/
├── driver
│   ├── include
│   └── src
├── serialization
│   ├── include
│   ├── platform
│   └── src
└── sl
    ├── include
    ├── lib
    ├── open
    │   └── src
    └── platform
```

**Comment by pkwiek** [comment]
Are we sticking to `driver` or switching to proposed sometime ago `kernel`?

**Comment by hubertmis** [comment]
> Are we sticking to `driver` or switching to proposed sometime ago `kernel`?

`kernel` would be technically more appropriate, however the name of the bundled solution is `driver`.
When an user is going to use this repository and integrate it to something, one would be confused if they should use `serialization`, `sl`, or `kernel`. Using `driver` instead of `kernel` to name this directory could lead such user into right direction what to use.

Alternatively, we could use `kernel` and good `README` describing that it should be used. But does everyone read `README`?

**Comment by jciupis** [comment]
> Shall we target something like this?

**Comment by czeslawmakarski** [comment]
I would vote for `driver` instead of `kernel`. Users are more used to it, and furthermore, `kernel` in RTOS context means something very very different.

I wonder if we could paste `sl/` to the .gitignore so that we don't mistakenly contribute to the SL out of this repo (not out of SL repo).

**Comment by jciupis** [comment]
I also think that `driver` is the better choice. It's already the coined term for our 802.15.4 solution and with the separation of features between `sl` and `serialization`, what's left of the whole bundle is pretty close to being an actual `driver`. Furthermore `kernel` is pretty misleading when one is to consider RTOS context, just like @czeslawmakarski mentioned.

**Comment by JakubPegza** [comment]
> > Are we sticking to `driver` or switching to proposed sometime ago `kernel`?
> 
> `kernel` would be technically more appropriate, however the name of the bundled solution is `driver`.

> I would vote for `driver` instead of `kernel`. Users are more used to it, and furthermore, `kernel` in RTOS context means something very very different.

Maybe we should consider using other name than 'kernel', but I think differentiating between component and bundled solution may be less confusing. I think (but I might be wrong here) that 'driver' is usually referring to this bundled solution, naming the component 'driver' may leave an impression that it's a standalone driver solution, and not just the core part, heavily reliant on sl (open or closed) and serialization (for nRF53)

**Comment by pkwiek** [comment]
From `driver` and `kernel` I would choose the first one, because after seeking for driver and finding `kernel` I would be somewhat confused and eventually read `README` (@hubertmis :D). @JakubPegza You have a point. This a good opportunity to think of a name, however I am unsure if this is worth the effort.

**Comment by pkwiek** [comment]
> Shall we target something like this?
> 
> ```
> sdk-802154/
> ├── driver
> │   ├── include
> │   └── src
> ├── serialization
> │   ├── include
> │   ├── platform
> │   └── src
> └── sl
>     ├── include
>     ├── lib
>     ├── open
>     │   └── src
>     └── platform
> ```

Assuming above structure. What are your thoughts on moving all platform files from driver to SL? I am aware that random platform is used only by the kernel.

**Comment by jciupis** [comment]
@pkwiek I'm not convinced. What would be the advantage of that approach over each component having its own platform? Probably it'd make sense if `serialization` didn't have its platform and moving the driver platform files to SL would equal having a single placeholder for all platform files. However, that's not the case, so in my opinion it' clearer if the components are symmetrical.

**Review by jciupis — APPROVED** [review]

File: driver/src/platform/random/nrf_802154_random.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (hubertmis) [inline driver/src/platform/random/nrf_802154_random.h]: I would move `platform` directory one level above, like we discussed in PR

File: sl/include/platform/clock/nrf_802154_clock.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
+/*
```
Comment (hubertmis) [inline sl/include/platform/clock/nrf_802154_clock.h]: Like in `driver/`, I would move platform one level above to clearly separate it from headers and implementation

Comment (hubertmis) [inline sl/include/platform/clock/nrf_802154_clock.h]: And we should decide on one approach:
`sl/platform/include/nrf_802154_clock.h` or `sl/platform/clock/nrf_802154_clock.h`. Right now the `driver` one differs from the `sl` one.

**Review by hubertmis — CHANGES_REQUESTED** [review]
It looks SPDX's are not yet applied. We should discuss if we want to apply them now, while creating repo, or port later with PR. Both are OK for me, just we need to think what would be more efficient.

Also, I've found small inconsistency in `platform` directory placement across modules. I think we should unify it.

File: sl/include/platform/clock/nrf_802154_clock.h (hunk 2 above)
Comment (pkwiek) [reply sl/include/platform/clock/nrf_802154_clock.h]: I am in favor of `sl/platform/include/nrf_802154_clock.h`. IMO it seems cleaner when includes are separated from sources. Moreover for CMake we would get `*include_directories(<dir_where_only_includes_exist>)`.

**Comment by pkwiek** [comment]
Licenses are updated.
Are there any other opinions about the matter below?
> And we should decide on one approach:
> `sl/platform/include/nrf_802154_clock.h` or `sl/platform/clock/nrf_802154_clock.h`. Right now the `driver` one differs from the `sl` one.

File: serialization/src/include/nrf_802154_spinel_backend.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (hubertmis) [inline serialization/src/include/nrf_802154_spinel_backend.h]: Following directory scheme for platform in `driver` and `sl` I think this file (and other platform files for serialization) should be moved to `serialization/platform/include`

**Review by hubertmis — CHANGES_REQUESTED** [review]
Look very good!
Last missing item is alignment of directory structure for serialization platform API with platform APIs of other modules.

**Comment by e-rk** [comment]
One thing I think should be kept in mind is the naming conflict between the radio driver public headers and serialization public headers. For example the inclusion of `nrf_802154_config.h` is ambiguous on nRF5340 network core, where both radio driver and serialization are used. Whether serialization or radio driver variant of the file gets included depends on the order of the include paths passed as an argument to the compiler.

**Comment by hubertmis** [comment]
> One thing I think should be kept in mind is the naming conflict between the radio driver public headers and serialization public headers

Good point. I think we should clearly distinguish which headers apply to host and which to the remote build. Is it that only platform API applies to remote, while both platform and public high level (driver-like) headers apply to host?

**Comment by e-rk** [comment]
@hubertmis Not sure if I understand correctly, but I think high-level API applies to both remote and host. The remote must see the high level API to communicate with the radio hardware, while the host must use the high level API to communicate with the remote.

I think the best solution would be to keep only one version of `nrf_802154.h` which does not contain any API that must be called by the platform, like `nrf_802154_radio_irq_handler`. That header would contain only the high-level API and would be used for both serialization and the driver.

Now about the `nrf_802154_config.h`, assuming that the `nrf_802154.h` contains only high level API, does the file really need to be exposed to the user? Now that we have run-time capabilities I think we can do away with this file and put it somewhere more hidden.

**Comment by hubertmis** [comment]
We have several API layers. Bottom-up:
1. Radio driver platform (used by radio driver) 
2. Radio driver public API (used by serialization module on remote core)
3. Serialization platform (used by remote and host parts of serialization module)
4. Serialization high level API, compatible with most Radio Driver calls (used by shim layer in host core).

I don't think we should mix serialization high level API with Radio Driver public API even though they should remain mostly compatible.

However, I agree we should clearly distinguish which headers should be used in host build, which in remote build an which in both. I understand nrf_802154.h from radio driver should be used in remote or single soc build, while nrf_802154.h from serialization only in serialization host build. This way there should be no problem with name conflicts. We shall never use both headers in the same build

**Comment by pkwiek** [comment]
> Now about the `nrf_802154_config.h`, assuming that the `nrf_802154.h` contains only high level API, does the file really need to be exposed to the user? Now that we have run-time capabilities I think we can do away with this file.

@e-rk, `nrf_802154_config.h` contains macros that can be used by a higher layer that depends on specific configuration options. @hubertmis, is my thinking correct?

>I don't think we should mix serialization high level API with Radio Driver public API even though they should remain mostly compatible.

+1

>However, I agree we should clearly distinguish which headers should be used in host build, which in remote build an which in both. I understand nrf_802154.h from radio driver should be used in remote or single soc build, while nrf_802154.h from serialization only in serialization host build. This way there should be no problem with name conflicts. We shall never use both headers in the same build

+1 and I am on it.

**Comment by hubertmis** [comment]
> @hubertmis, is my thinking correct?

That's also my understanding. Moreover, this file may be modified (directly or by a build system) by an user to tweak the driver for given application. Because of this I think it should be kept with public API (or with platform API), where it can be easily found by users.

**Comment by pkwiek** [comment]
Serialization platform is aligned. Please inform me if I missed something.
Serialization public API is divided into `host` part (containing dirver-like headers, included only for serialization host) and `serialization` part (always included).

**Review by hubertmis — APPROVED** [review]
Looks perfect now! Thanks

**Review by e-rk — APPROVED** [review]

**Review by JakubPegza — APPROVED** [review]

**Comment by JakubPegza** [comment]
What about moving platform headers in each module from platform/include to include/platform? This way we get a structure like this:
.
├── include
│   ├── normal_headers.h
│   └── platform
│       └── platform_headers.h
├── platform
│   └── platform_1_dir
│       ├── platform_1_src_1.c
│       └── platform_1_src_2.c
└── src
    ├── internal_headers.h
    └── sources.c
Main advantage is including headers by:

#include "some_header.h"
#include "platform/some_platform_header.h"

Instead of either (depending on CMakeLists contruction):

#include "some_header.h"
#include "platform/include/some_platform_header.h"
or
#include "some_header.h"
#include "some_platform_header.h"

Seems cleaner to me.
Opinions, @hubertmis @pkwiek? Are there any disadvantages to this approach?

**Comment by pkwiek** [comment]
@JakubPegza Seems cleaner in code due to `platform/header.h` convention. However there is still some mixup, because then we will have platform separated only in terms of sources (headers are next to public module headers). In that case platform sources could end up in `src` directories to make it symmetrical. And I do not know if this was the goal. Please, correct me if I got something wrong.

**Review by czeslawmakarski — APPROVED** [review]

**Pull request merged**